## PARCIAL # 1 ALGEBRA PARA CIENCIA DE DATOS

In [ ]:
#Jhon Hernandez Slait
#jjhernans1@eafit.edu.co
#Algebra para ciencia de datos

## Parte 1 - La distania de hamming entre dos valores booleanos de longitud n es el # de entradas en que difieren

a) Demuestre que efectivamente es una distancia


In [56]:
import itertools


esta función verifica que las dos secuencias tengan la misma longitud, luego compara las secuencias elemento por elemento y cuenta las diferencias. Finalmente, devuelve la distancia de Hamming, que es igual al número de diferencias entre las secuencias.

In [57]:
def hamming_distance(seq1, seq2):
    if len(seq1) != len(seq2):
        raise ValueError("Las secuencias deben tener la misma longitud para calcular la distancia de Hamming")
    
    distance = 0
    
    for i in range(len(seq1)):
        if seq1[i] != seq2[i]:
            distance += 1
    
    return distance


In [58]:
seq_a = "1010101"
seq_b = "1101010"
seq_c = "0010101"


In [62]:
distance_ab = hamming_distance(seq_a, seq_b)
distance_ac = hamming_distance(seq_a, seq_c)
distance_bc = hamming_distance(seq_b, seq_c)
#Esto calculará la distancia de Hamming entre las secuencias seq_a y seq_b, entre seq_a y seq_c, y entre seq_b y seq_c. 

In [60]:
# Positividad
assert distance_ab >= 0
assert distance_ac >= 0
assert distance_bc >= 0

# Simetría
assert distance_ab == hamming_distance(seq_b, seq_a)
assert distance_ac == hamming_distance(seq_c, seq_a)
assert distance_bc == hamming_distance(seq_c, seq_b)

# Desigualdad triangular
assert distance_ab <= distance_ac + distance_bc


Positividad: aqui verifico que las distancias de Hamming calculadas (distance_ab, distance_ac y distance_bc) sean números no negativos. Esto asegura que la distancia de Hamming siempre sea un valor positivo o igual a cero, ya que no puede haber una distancia negativa.

Simetría: Aca verifico que la distancia de Hamming entre dos secuencias (seq_a y seq_b, seq_a y seq_c, seq_b y seq_c) sea igual cuando se intercambian las secuencias.

Desigualdad Triangular:aqui valido la propiedad de la desigualdad triangular. Esto significa que la distancia de Hamming entre dos secuencias (seq_a y seq_b) debe ser menor o igual a la suma de las distancias de Hamming entre la primera secuencia y una tercera secuencia (seq_c) y la distancia de Hamming entre la segunda secuencia y la misma tercera secuencia. En otras palabras, se comprueba que distance_ab sea menor o igual a distance_ac + distance_bc. Esta propiedad es importante para asegurar que la distancia de Hamming cumple con las propiedades métricas.


In [70]:
print("La distancia de Hamming cumple con las tres propiedades clave de una métrica.")

La distancia de Hamming cumple con las tres propiedades clave de una métrica.


Si alguna de estas afirmaciones falla (es decir, si alguna propiedad no se cumple), se producirá un error que indicará que la distancia de Hamming no satisface alguna de las propiedades clave de una métrica. Si todas las afirmaciones pasan sin errores, significa que la distancia de Hamming cumple con las propiedades requeridas para ser considerada una métrica válida.

## Parte 2: Encuentre un conjunto de datos con al menos 5 variables booleanas / 4 de entradas / 1 de objetivo.

In [33]:
import requests
import pandas as pd
from io import StringIO

# se especifica la URL del archivo CSV en GitHub.
github_url = 'https://raw.githubusercontent.com/Johnslait/algebraciencia/main/bank.csv'

# Realizo una solicitud GET para obtener el contenido del archivo CSV
response = requests.get(github_url)

# Verifico si la solicitud fue exitosa (código de estado 200)
if response.status_code == 200:
    # Lee el contenido del archivo en un DataFrame de Pandas
    contenido = StringIO(response.text)
    df = pd.read_csv(contenido)
    
    # Ahora, df contiene los datos del archivo CSV desde GitHub
    # Ya puedo realizar operaciones de análisis de datos en df
    
    # Por ejemplo, mostrar las primeras filas del DataFrame
    print(df.head())
else:
    print("No se pudo obtener el archivo. Código de estado:", response.status_code)


   age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes  


Ahora se crea una nueva columna llamada 'new_marital' en el DataFrame df aplicando una función personalizada a la columna 'marital'. Esta función transforma los valores 'single' y 'divorced' en 'no casado' y los demás valores en 'casado'.

Se crea un diccionario de mapeo llamado mapeo_booleano que asigna 'casado' a True y 'no casado' a False.

Se aplica este mapeo a la columna 'new_marital' del DataFrame, convirtiendo así los valores 'casado' en True y 'no casado' en False.

In [54]:

# Función personalizada para la transformación
def transformar_estado_civil(marital):

    if marital in ['single', 'divorced']:
        return 'no casado'
    else:
        return 'casado'

# Aplicar la función a la columna 'marital' y crear la nueva columna 'casado'
df['new_marital'] = df['marital'].apply(transformar_estado_civil)

# Crear un diccionario de mapeo para convertir 'casado' y 'no casado' a booleanos
mapeo_booleano = {'casado': True, 'no casado': False}

# Aplicar el mapeo a la columna 'new_marital' para convertirla a booleana
df['new_marital'] = df['new_marital'].map(mapeo_booleano)

# Mostrar el DataFrame resultante
print(df['new_marital'])


0         True
1         True
2         True
3         True
4         True
         ...  
11157    False
11158     True
11159    False
11160     True
11161     True
Name: new_marital, Length: 11162, dtype: bool


Esta parte del código crea un nuevo DataFrame llamado `df_bool` que contiene las mismas columnas que se encuentran en el DataFrame original `df`, pero solo selecciono y copio aquellas columnas que se encuentran en mi lista `columnas_a_convertir`. Las columnas que he seleccionado son: 'housing', 'loan', 'default' y 'deposit'.

Entonces, básicamente, estoy creando un DataFrame nuevo que es una copia de un subconjunto de columnas del DataFrame original. Este nuevo DataFrame `df_bool` contendrá únicamente las columnas que planeo convertir en variables booleanas. Esto me permite trabajar de manera más específica en la conversión de esas columnas sin afectar el DataFrame original.

In [35]:
# Crea un nuevo DataFrame con las columnas que deseas convertir en booleanas
columnas_a_convertir = ['housing', 'loan', 'default', 'deposit']
df_bool = df[columnas_a_convertir].copy()



In [36]:
# Utiliza el método map para convertir las columnas en booleanas
for columna in columnas_a_convertir:
    df_bool[columna] = df_bool[columna].map({'yes': True, 'no': False})



In [37]:
# Concatena el nuevo DataFrame con el DataFrame original
df_nuevo = pd.concat([df['new_marital'],df_bool], axis=1)
df_nuevo


new_marital  housing   loan  default  deposit
0             True     True  False    False     True
1             True    False  False    False     True
2             True     True  False    False     True
3             True     True  False    False     True
4             True    False  False    False     True
...            ...      ...    ...      ...      ...
11157        False     True  False    False    False
11158         True    False  False    False    False
11159        False    False  False    False    False
11160         True    False   True    False    False
11161         True    False  False    False    False

[11162 rows x 5 columns]

In [64]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, hamming_loss
from sklearn.neighbors import KNeighborsClassifier
from scipy.spatial.distance import hamming
# Dividir el conjunto de datos en características (X) y variable objetivo (y)
X = df_nuevo[['housing','loan','default','new_marital']]
y = df_nuevo['deposit']


In [65]:
# Dividir el conjunto de datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [67]:
# Entrenar un clasificador - DecisionTree para la prueba
clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = clf.predict(X_test)

# Evaluar el desempeño del clasificador
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
hamming_distance = hamming_loss(y_test, y_pred)


In [69]:
# Suponiendo que tienes dos vectores booleanos o binarios, y_test y y_pred
# Convertir a listas si son objetos de pandas
list_y_test = y_test.tolist()
list_y_pred = y_pred.tolist()

# Calcular la distancia de Hamming
distance = hamming(list_y_test, list_y_pred)
print(f"La distancia de Hamming entre y_test y y_pred es: {distance}")


# Si deseas la distancia en términos de número de etiquetas mal clasificadas
num_labels = len(list_y_test)
num_misclassified = int(distance * num_labels)
print(f"Número de etiquetas mal clasificadas: {num_misclassified}")
print('Numero total de etiquetas: ', num_labels)
# Imprimir las métricas de desempeño
print(f"Exactitud (Accuracy): {accuracy}")
print(f"Precisión (Precision): {precision}")
print(f"Recuerdo (Recall): {recall}")
print(f"Puntuación F1 (F1 Score): {f1}")

La distancia de Hamming entre y_test y y_pred es: 0.3631885356023287
Número de etiquetas mal clasificadas: 811
Numero total de etiquetas:  2233
Exactitud (Accuracy): 0.6368114643976713
Precisión (Precision): 0.6230769230769231
Recuerdo (Recall): 0.6073102155576382
Puntuación F1 (F1 Score): 0.6150925486473658



1. Exactitud (Accuracy): 0.6368
   - El clasificador tiene una exactitud del 63.68%, lo que significa que aproximadamente el 63.68% de las predicciones hechas por el modelo son correctas.

2. Precisión (Precision): 0.623
   - el clasificador tiene una precisión del 62.31%, lo que significa que aproximadamente el 62.31% de las instancias clasificadas como positivas son realmente positivas.

3. Recuerdo (Recall): 0.6073
   - el clasificador tiene un recuerdo del 60.73%, lo que significa que aproximadamente el 60.73% de todas las instancias positivas fueron identificadas correctamente.

4. Puntuación F1 (F1 Score): el clasificador tiene una puntuación F1 de 0.6151.

5. Distancia de Hamming de desempeño: 0.5186
la distancia de Hamming mide la diferencia entre las etiquetas reales y las etiquetas predichas en términos de cuántos elementos son diferentes. En este caso, la distancia de Hamming promedio entre las etiquetas reales y predichas es 0.5186.
